File: src/problem018/solution018.hs
import Data.List (zipWith)

main = do
    content <- readFile "triangle.txt"
    let rows = map (\line -> [read x::Int | x <- words line]) $ lines content
    print . head . foldr1 (\r1 r2 -> zipWith (+) r1 (reduce r2)) $ rows

reduce row = zipWith max row (tail row)

File: src/problem018/solution018.py
import functools

def row_reduce(row):
    return [max(x1, x2) for (x1, x2) in zip(row, row[1:])]

def foldl_rows(row1, row2):
    return [x1 + x2 for (x1, x2) in zip(row2, row_reduce(row1))]

def main():
    rows = [map(int, line.split()) for line in open('triangle.txt').readlines()]
    return max(functools.reduce(foldl_rows, rows[::-1]))

if __name__ == '__main__':
    print(main())
